File: sections/100_grandmaster_finale/problems/12_parametric_quota_cut/solution.py
import collections
import sys

INF = 10**30


class Dinic:
    def __init__(self, n):
        self.graph = [[] for _ in range(n)]
        self.level = [-1] * n
        self.next_edge = [0] * n

    def add_edge(self, u, v, capacity):
        self.graph[u].append([v, len(self.graph[v]), capacity])
        self.graph[v].append([u, len(self.graph[u]) - 1, 0])

    def max_flow(self, source, sink):
        total = 0
        n = len(self.graph)
        while True:
            self.level[:] = [-1] * n
            self.level[source] = 0
            queue = collections.deque([source])
            while queue:
                u = queue.popleft()
                for v, _, capacity in self.graph[u]:
                    if capacity and self.level[v] == -1:
                        self.level[v] = self.level[u] + 1
                        queue.append(v)
            if self.level[sink] == -1:
                return total
            self.next_edge[:] = [0] * n

            def send(u, pushed):
                if u == sink:
                    return pushed
                while self.next_edge[u] < len(self.graph[u]):
                    edge_index = self.next_edge[u]
                    edge = self.graph[u][edge_index]
                    v, reverse, capacity = edge
                    if capacity and self.level[v] == self.level[u] + 1:
                        sent = send(v, min(pushed, capacity))
                        if sent:
                            edge[2] -= sent
                            self.graph[v][reverse][2] += sent
                            return sent
                    self.next_edge[u] += 1
                return 0

            while True:
                pushed = send(source, INF)
                if not pushed:
                    break
                total += pushed


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    index = 0
    n, m, target = data[index:index + 3]
    index += 3
    profits = data[index:index + n]
    index += n
    dependencies = []
    for _ in range(m):
        dependencies.append((data[index] - 1, data[index + 1] - 1))
        index += 2

    def solve(penalty):
        source, sink = n, n + 1
        flow = Dinic(n + 2)
        for vertex, profit in enumerate(profits):
            adjusted = profit - penalty
            if adjusted > 0:
                flow.add_edge(source, vertex, adjusted)
            elif adjusted < 0:
                flow.add_edge(vertex, sink, -adjusted)
        for u, v in dependencies:
            flow.add_edge(u, v, INF)
        flow.max_flow(source, sink)
        reachable = [False] * (n + 2)
        reachable[source] = True
        queue = [source]
        for u in queue:
            for v, _, capacity in flow.graph[u]:
                if capacity and not reachable[v]:
                    reachable[v] = True
                    queue.append(v)
        size = 0
        original_profit = 0
        for vertex in range(n):
            if reachable[vertex]:
                size += 1
                original_profit += profits[vertex]
        return size, original_profit

    low, high = -1_000_000_001, 1_000_000_001
    while low < high:
        middle = (low + high) // 2
        if solve(middle)[0] <= target:
            high = middle
        else:
            low = middle + 1
    size, answer = solve(low)
    if size != target:
        raise RuntimeError("input violates the promised attainable cardinality")
    print(answer)


if __name__ == "__main__":
    sys.setrecursionlimit(1_000_000)
    main()

File: sections/100_grandmaster_finale/problems/12_parametric_quota_cut/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using int64 = long long;

constexpr int64 INF = 4'000'000'000'000'000'000LL;

struct Dinic {
    struct Edge { int to, reverse; int64 capacity; };
    int n;
    vector<vector<Edge>> graph;
    vector<int> level, next_edge;

    explicit Dinic(int n) : n(n), graph(n), level(n), next_edge(n) {}

    void add_edge(int u, int v, int64 capacity) {
        graph[u].push_back({v, static_cast<int>(graph[v].size()), capacity});
        graph[v].push_back({u, static_cast<int>(graph[u].size()) - 1, 0});
    }

    bool bfs(int source, int sink) {
        fill(level.begin(), level.end(), -1);
        queue<int> queue;
        level[source] = 0;
        queue.push(source);
        while (!queue.empty()) {
            int u = queue.front();
            queue.pop();
            for (const Edge &edge : graph[u])
                if (edge.capacity > 0 && level[edge.to] == -1) {
                    level[edge.to] = level[u] + 1;
                    queue.push(edge.to);
                }
        }
        return level[sink] != -1;
    }

    int64 dfs(int u, int sink, int64 pushed) {
        if (u == sink) return pushed;
        for (int &index = next_edge[u]; index < static_cast<int>(graph[u].size()); ++index) {
            Edge &edge = graph[u][index];
            if (edge.capacity == 0 || level[edge.to] != level[u] + 1) continue;
            int64 sent = dfs(edge.to, sink, min(pushed, edge.capacity));
            if (sent == 0) continue;
            edge.capacity -= sent;
            graph[edge.to][edge.reverse].capacity += sent;
            return sent;
        }
        return 0;
    }

    int64 max_flow(int source, int sink) {
        int64 result = 0;
        while (bfs(source, sink)) {
            fill(next_edge.begin(), next_edge.end(), 0);
            while (int64 pushed = dfs(source, sink, INF)) result += pushed;
        }
        return result;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m, target;
    cin >> n >> m >> target;
    vector<int64> profit(n);
    for (int64 &x : profit) cin >> x;
    vector<pair<int, int>> dependencies(m);
    for (auto &[u, v] : dependencies) {
        cin >> u >> v;
        --u;
        --v;
    }

    auto solve = [&](int64 penalty) {
        int source = n, sink = n + 1;
        Dinic flow(n + 2);
        for (int vertex = 0; vertex < n; ++vertex) {
            int64 adjusted = profit[vertex] - penalty;
            if (adjusted > 0) flow.add_edge(source, vertex, adjusted);
            else if (adjusted < 0) flow.add_edge(vertex, sink, -adjusted);
        }
        for (auto [u, v] : dependencies) flow.add_edge(u, v, INF);
        flow.max_flow(source, sink);
        vector<bool> reachable(n + 2);
        queue<int> queue;
        reachable[source] = true;
        queue.push(source);
        while (!queue.empty()) {
            int u = queue.front();
            queue.pop();
            for (const auto &edge : flow.graph[u])
                if (edge.capacity > 0 && !reachable[edge.to]) {
                    reachable[edge.to] = true;
                    queue.push(edge.to);
                }
        }
        int size = 0;
        int64 original_profit = 0;
        for (int vertex = 0; vertex < n; ++vertex)
            if (reachable[vertex]) {
                ++size;
                original_profit += profit[vertex];
            }
        return pair{size, original_profit};
    };

    int64 low = -1'000'000'001LL, high = 1'000'000'001LL;
    while (low < high) {
        int64 middle = low + (high - low) / 2;
        if (solve(middle).first <= target) high = middle;
        else low = middle + 1;
    }
    auto [size, answer] = solve(low);
    if (size != target) {
        // The input promise guarantees this branch is unreachable.
        return 1;
    }
    cout << answer << '\n';
}
